src/app/content/utils/canonicalUrl.ts

```diff
@@ -14,29 +15,50 @@ export async function getCanonicalUrlParams(
   pageId: string,
   bookVersion?: string
 ) {
-  const bookDefaultMap = [[book.id, {}]] as Array<[string, ObjectLiteral<undefined>]>;
   const getBook = makeUnifiedBookLoader(archiveLoader, osWebLoader);
-  const bookVersionFromConfig = getBookVersionFromUUIDSync(book.id);
-  const canonicals = ([
-    ...(CANONICAL_MAP[book.id] || []),
-    ...(bookVersionFromConfig && bookVersion === bookVersionFromConfig.defaultVersion ? bookDefaultMap : []),
-    // use the current book as a last resort if it has the same version as in books config
-  ]).filter(([id]) => !!getBookVersionFromUUIDSync(id));
 
-  for (const [id, CANONICAL_PAGES_MAP] of canonicals) {
-    const version = assertDefined(
-      getBookVersionFromUUIDSync(id),
-      `We've already filtered out books that are not in the BOOK configuration`
-    ).defaultVersion;
-    const canonicalBook = book.id === id  && hasOSWebData(book) ? book : await getBook(id, version);
-    const mappedPageId = CANONICAL_PAGES_MAP[pageId] || pageId;
-    const treeSection = findArchiveTreeNodeById(canonicalBook.tree, mappedPageId);
+  const getCanonicalMap = (bookId: string) => {
+    const bookDefaultMap = [[bookId, {}]] as Array<[string, ObjectLiteral<undefined>]>;
+    const bookVersionFromConfig = getBookVersionFromUUIDSync(bookId);
+    return ([
+      ...(CANONICAL_MAP[bookId] || []),
+      ...(bookVersionFromConfig && bookVersion === bookVersionFromConfig.defaultVersion ? bookDefaultMap : []),
+      // use the current book as a last resort if it has the same version as in books config
+    ]).filter(([id]) => !!getBookVersionFromUUIDSync(id));
+};
 
-    if (treeSection) {
-      const pageInBook = assertDefined(treeSection.slug, 'Expected page to have slug.');
-      return {book: {slug: canonicalBook.slug}, page: {slug: pageInBook}};
+  let canonicalMap = getCanonicalMap(book.id);
+  let canonicalPageId = pageId;
+  let done = false;
+  let canonicalBook;
+  let treeSection;
+
+  while (!done) {
+    for (const [id, CANONICAL_PAGES_MAP] of canonicalMap) {
+      const version = assertDefined(
+        getBookVersionFromUUIDSync(id),
+        `We've already filtered out books that are not in the BOOK configuration`
+      ).defaultVersion;
+      const useCurrentBookAsCanonical = book.id === id  && hasOSWebData(book);
+      canonicalBook = useCurrentBookAsCanonical ? book : await getBook(id, version);
+      canonicalPageId = CANONICAL_PAGES_MAP[canonicalPageId] || canonicalPageId;
+      treeSection = findArchiveTreeNodeById(canonicalBook.tree, canonicalPageId);
+
+      if (!useCurrentBookAsCanonical) {
+        const newMap = getCanonicalMap(canonicalBook.id);
+        // stop when we run out of canonical maps to check
+        done = !newMap.length || isEqual(canonicalMap, newMap);
+        canonicalMap = newMap;
```

Comment:
It's possible there would be a cycle in the CANONICAL_MAP and this would run forever, alternating between them. Not sure how important it is to prevent this.

Reply:
Per today's conversation it sounds like that would be the expected behavior.

Reply:
Now this throws if it's about to run through a map that's already been checked.